teps/0124-distributed-tracing-for-tasks-and-pipelines.md

```diff
@@ -0,0 +1,98 @@
+---
+status: proposed
+title: Distributed tracing for Tasks and Pipelines
+creation-date: '2022-09-30'
+last-updated: '2022-09-30'
+authors:
+- '@kmjayadeep'
+---
+
+# TEP-0124: Distributed tracing for Tasks and Pipelines
+
+<!-- toc -->
+- [Summary](#summary)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [PipelineRun controller](#pipelinerun-controller)
+  - [TaskRun controller](#taskrun-controller)
+  - [Test Plan](#test-plan)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+With distributed tracing, we can track the time taken by each action in the pipeline like reconciling logic, fetching resources, pulling images etc.
+This allows the developers to improve the reconciliation logic and also allow end users to monitor and optimize the pipelines.
+
+### Goals
+
+* Implementation of opentelemetry tracing with Jaeger
+* Instrumentation of pipelinerun and taskrun reconciliation logic
+* Able to visualize pipeline and task reconciliation steps in jaeger
+
+### Non-Goals
+
+* Instrumentation of sidecars and initcontainers
+* Support for more tracing backends
+* Instrumentation of individual steps in each task
+* Adding events to each span to indicate what is happening inside each
+  method. This is an improvement task, and it can be done later once the
+  basic setup for tracing is in place. The scope of this proposal is
+  only to get the plumbing done by covering only the method boundaries.
+
+### Use Cases
+
+Pipeline and Task User:
+* I would like to understand the duration of each task in my pipeline so that I can optimize the slow taks to improve the pipeline execution speed
+
+Tekton Developer:
+* I would like to understand the duration of each reconciliation step, so that I can optimize the code to improve reconciliation performance
+* When the pipelines are failing due to a bug, I would like to understand which reconciliation logic caused the issue so that I can easily fix the problem
+
+### Requirements
+
+* Trace all functions in the PipelineRun controller
+* Trace all functions in the TaskRun controller
+* Support Jaeger backend
+* Propagate traces so that subsequent reconciles of the same resource belong to the same trace
+* Propagate traces so that reconciles of a resource owned by a parent resource belong a parent span from the parent resource
+* Reconcile of different resources must belong to separate traces
+
+## Proposal
```

Comment:
A small architecture diagram of how jaeger integrates with tekton

Reply:
I'm not sure how to come up with an architecture diagram for this. There are only two relevant components, Tekton operator and Jaeger. Tekton operator just records the spans during reconcile process and flush them to jaeger on specific intervals.

Reply:
Sounds good. Just wanted to make sure if there was anything out of the ordinary being introduced.

Reply:
Thanks @pxp928  - a diagram would be nice indeed.
We could add a diagram in the next PR when we propose this as implementable if that's ok